### PR TITLE
Add initial membership view implementation

### PR DESCRIPTION
**Pull request by LucioFranco**

### CODE REVIEW

File: src/error.rs [hunk 1]
```diff
@@ -9,15 +9,23 @@ pub struct Error {
     source: Option<Source>,
 }
 
-#[derive(Debug)]
+#[derive(Debug, Eq, PartialEq)]
 pub(crate) enum ErrorKind {
```
Comment (bIgBV) [inline src/error.rs:13]: one thing to think about is whether or not we want to split error kind into different enums as the number of errors grows.

File: src/membership/ring.rs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+impl<T: AsRef<[u8]>> Ring<T> {
+    pub fn new(seed: u64) -> Self {
+        Self {
+            seed,
+            set: BTreeSet::new(),
+        }
+    }
+
+    pub fn contains(&self, key: T) -> bool {
+        // TODO(lucio): Remove this clone since we only need a `AsRef<[u8]>` to
```
Comment (bIgBV) [inline src/membership/ring.rs:25]: This TODO is done?

File: src/membership/ring.rs [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        self.set.insert(seeded_key)
+    }
+
+    pub fn get(&mut self, key: T) -> Option<&T> {
+        let seeded_key = SeededKey::new(key, self.seed);
+        self.set.get(&seeded_key).map(|v| &v.key)
+    }
+
+    pub fn remove(&mut self, key: T) -> bool {
+        // TODO(lucio): Remove this clone since we only need a `AsRef<[u8]>` to
```
Comment (bIgBV) [inline src/membership/ring.rs:42]: Same as above?

File: src/membership/ring.rs [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        // do the `Ord` impl.
+        let seeded_key = SeededKey::new(key, self.seed);
+        self.set.remove(&seeded_key)
+    }
+
+    pub fn higher(&self, key: T) -> Option<&T> {
+        let seeded_key = SeededKey::new(key, self.seed);
+        let range = (Bound::Excluded(seeded_key), Bound::Unbounded);
```
Comment (bIgBV) [inline src/membership/ring.rs:50]: Oooh this is a neat way to create a range.

File: src/membership/view.rs [hunk 5]
```diff
@@ -1,15 +1,295 @@
...
+            should_update_configuration_id: true,
+        }
+    }
+
+    pub fn ring_add(&mut self, node: Endpoint, node_id: NodeId) -> Result<()> {
+        if self.is_node_present(&node_id) {
+            return Err(Error::new_uuid_already_seen());
+        }
+
+        if self.rings[0].contains(node.clone()) {
```
Comment (bIgBV) [inline src/membership/view.rs:50]: So I'm going to make a separate comment about this, but I think it's important to document why checking only the first ring is fine. Reading the paper would tell you that all rings have the same nodes, so checking one is enough. But you need to remember this context.

File: src/membership/view.rs [hunk 6]
```diff
@@ -1,15 +1,295 @@
...
+};
+use std::collections::HashSet;
+
+type Ring = crate::membership::ring::Ring<Endpoint>;
+
 #[derive(Debug, Clone)]
 pub struct View {
-    configuration: Configuration,
+    k: i32,
+    rings: Vec<Ring>,
```
Comment (bIgBV) [inline src/membership/view.rs:12]: One day when rust gets specialization, we can make this an array of size `k`...one day...

File: src/membership/view.rs [hunk 7]
```diff
@@ -1,15 +1,295 @@
...
+            return Err(Error::new_node_not_in_ring());
+        }
+
+        if self.rings[0].len() <= 1 {
+            return Ok(Vec::new());
+        }
+
+        let mut observers = Vec::new();
+
+        for ring in &self.rings {
```
Comment (bIgBV) [inline src/membership/view.rs:90]: Again maybe a comment as to why this logic extracts the observers for a given node from the rings.

**Review by bIgBV — APPROVED** [review]
Looks good to me. But I think we need to add the crate level warning for adding docs. This is a major module, with a lot of logic which isn't obvious at first glance. I'd be more comfortable if we at least _tried_ to add some docs 😅

File: src/error.rs (hunk 1 above)
Comment (LucioFranco) [reply src/error.rs:13]: Probably

File: src/membership/ring.rs (hunk 2 above)
Comment (LucioFranco) [reply src/membership/ring.rs:25]: no gonna have to leave it in couldn't find a way to get the ref to work :(

File: src/membership/view.rs (hunk 5 above)
Comment (LucioFranco) [reply src/membership/view.rs:50]: The reason this is safe is because when you create the view `K > 1` meaning that there is atleast one ring.

**Comment by LucioFranco** [comment]
I'm gonna merge this then PR docs and comments

**Pull request merged**